src/UriSetter.c

```diff
@@ -0,0 +1,220 @@
+ * SERVICES; LOSS OF USE, DATA,  OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT  LIABILITY,  OR  TORT (INCLUDING  NEGLIGENCE  OR  OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file UriSetter.c
+ * Holds the RFC 3986 %URI setter implementation.
+ * NOTE: This source file includes itself twice.
+ */
+
+/* What encodings are enabled? */
+#include <uriparser/UriDefsConfig.h>
+#if (!defined(URI_PASS_ANSI) && !defined(URI_PASS_UNICODE))
+/* Include SELF twice */
+# ifdef URI_ENABLE_ANSI
+#  define URI_PASS_ANSI 1
+#  include "UriSetter.c"
+#  undef URI_PASS_ANSI
+# endif
+# ifdef URI_ENABLE_UNICODE
+#  define URI_PASS_UNICODE 1
+#  include "UriSetter.c"
+#  undef URI_PASS_UNICODE
+# endif
+#else
+# ifdef URI_PASS_ANSI
+#  include <uriparser/UriDefsAnsi.h>
+# else
+#  include <uriparser/UriDefsUnicode.h>
+#  include <wchar.h>
+# endif
+
+
+
+#ifndef URI_DOXYGEN
+# include <uriparser/Uri.h>
+# include "UriNormalizeBase.h"
+# include "UriCommon.h"
+# include "UriMemory.h"
+#endif
+
+
+
+#include <assert.h>
+
+
+
+static int URI_FUNC(SetSchemeEngine)(URI_TYPE(Uri) * uri,
+		const URI_CHAR * first, const URI_CHAR * afterLast,
+		UriMemoryManager * memory);
+
+static URI_INLINE UriBool URI_FUNC(SetRange)(URI_TYPE(TextRange) * range,
+		const URI_CHAR * first, const URI_CHAR * afterLast,
+		UriBool owner, UriMemoryManager * memory);
+
+
+
+int URI_FUNC(SetSchemeEx)(URI_TYPE(Uri) * uri,
+		const URI_CHAR * first, const URI_CHAR * afterLast) {
+	return URI_FUNC(SetSchemeExMm)(uri, first, afterLast, NULL);
+}
+
+
+
+int URI_FUNC(SetSchemeExMm)(URI_TYPE(Uri) * uri,
+		const URI_CHAR * first, const URI_CHAR * afterLast,
+		UriMemoryManager * memory) {
+	URI_CHECK_MEMORY_MANAGER(memory);  /* may return */
+	return URI_FUNC(SetSchemeEngine)(uri, first, afterLast, memory);
+}
+
+
+
+int URI_FUNC(SetScheme)(URI_TYPE(Uri) * uri,
+		const URI_CHAR * first, const URI_CHAR * afterLast) {
+	return URI_FUNC(SetSchemeEx)(uri, first, afterLast);
+}
+
+
+
+static UriBool URI_FUNC(ValidateScheme)(const URI_CHAR * first,
+	const URI_CHAR * afterLast) {
+	const URI_CHAR * i;
+
+	switch (*first) {
+	case URI_SET_ALPHA:
+		break;
+	default:
+		return URI_FALSE;
+	}
+
+	for (i = first + 1; i < afterLast; i++) {
+		switch (*first) {
+		case _UT('.'):
+		case _UT('+'):
+		case _UT('-'):
+		case URI_SET_ALPHA:
+		case URI_SET_DIGIT:
+			continue;
+		default:
+			return URI_FALSE;
+		}
+	}
+
+	return URI_TRUE;
+}
+
+
+
+static int URI_FUNC(SetSchemeEngine)(URI_TYPE(Uri) * uri,
+		const URI_CHAR * first, const URI_CHAR * afterLast,
+		UriMemoryManager * memory) {
+	if (uri == NULL) {
+		return URI_ERROR_NULL;
+	}
+
+	if (first != NULL && URI_FUNC(ValidateScheme)(first, afterLast) == URI_FALSE) {
+		return URI_ERROR_SYNTAX;
+	}
+
+	if (URI_FUNC(SetRange)(&uri->scheme, first, afterLast, uri->owner, memory) == URI_FALSE) {
```

Comment:
If this succeeds with `uri->owner` set but the rest of the function fails somewhere below, the duped scheme memory would be leaked.